File: packages/audio-tap/android/src/main/java/com/clarionhq/audio_tap/AudioTapTypes.kt
package com.clarionhq.audio_tap

/**
 * Internal throwable. Mapped to `NativeAudioTapError` at the bridge layer
 * (`HybridClarionAudioTap`).
 */
internal class AudioTapError(
  val code: String,
  override val message: String,
  val recoverable: Boolean = false,
) : RuntimeException(message)

internal interface AudioTapCallbacks {
  fun onState(state: String)
  fun onFrame(
    pcm: ByteArray,
    timestamp: Double,
    frameIndex: Long,
    sampleRate: Int,
    channels: Int,
    bitsPerSample: Int,
  )
  fun onStats(
    uptimeMs: Double,
    framesEmitted: Long,
    framesDropped: Long,
    bufferFillPct: Double,
  )
  fun onError(error: AudioTapError)
}

File: packages/audio-tap/ios/AudioTapTypes.swift
import Foundation

/// Internal Swift-throwable. Mapped to `NativeAudioTapError` at the Nitro
/// bridge layer (`HybridClarionAudioTap`). We can't make
/// `NativeAudioTapError` directly `Error`-conformant — it's a C++ struct
/// typealias.
internal struct AudioTapErr: Error, CustomStringConvertible {
  let code: String
  let message: String
  let recoverable: Bool

  init(code: String, message: String, recoverable: Bool = false) {
    self.code = code
    self.message = message
    self.recoverable = recoverable
  }

  var description: String { "AudioTapErr(\(code)): \(message)" }

  func toNative() -> NativeAudioTapError {
    return NativeAudioTapError(code: code, message: message, recoverable: recoverable)
  }
}
